### ISSUE TITLE
AdBlocking stops silently

### ISSUE THREAD

**Issue by hrbrgr** [julian-klode/dns66#36]
Hi there, thanks for this great piece of software.
I might experience a bug, though.
Every once in a while, ads reappear on websites after a few hours.
A quick manual disconnect followed by a reconnect in your app solves the problem and ads immediately vanish when refreshing the page.

**Comment by FernandoMiguel** [comment]
What version of Android? 
Maybe it's a power optimization that is killing the app. 
Can you try adding it to Battery Optimisation 
![screenshot_20161028-070745](https://cloud.githubusercontent.com/assets/168366/19796583/611257d2-9cdd-11e6-83d2-49768a458976.png)

**Comment by hrbrgr** [comment]
Good morning,

I'm using a Samsung S7 Android 6.0.1 stock Rom.
I will try that, good guess.

**Comment by hrbrgr** [comment]
Hi there,
No more problems after disabling battery optimisation (doze) as well as "app power saving" in battery menu (app standby).

Thanks again.

**Comment by julian-klode** [comment]
That's great. I really only test this stuff on OnePlus devices (a 3 with OxygenOS 3.5, and a One running a Nougat Pure Nexus Rom build with pico gapps), so I can't know how other devices behave.

I wonder if we can notice this somehow and automatically setup again...

**Comment by hrbrgr** [comment]
You could probably link at least to the doze menu, like e.g. greenery does.
And maybe show a note before with instructions.

**Comment by llam15** [comment]
Hi, I'm running OxygenOS 3.2.7 on an OP3. I turned off battery optimization, but I still get the issue where it silently stops after awhile. Any ideas on how I can fix this?

**Comment by julian-klode** [comment]
Hmm, I don't see this on Oxygen OS 3.5. That really should not happen, that's why we show the persistent notification.

**Comment by llam15** [comment]
Yeah, the persistent notification will disappear, and when I open the app it says either "Stopping" or "Stopped"

**Comment by mhaas** [comment]
I am seeing the same issue. I do have the key symbol in my notification bar, but it will vanish every couple of minutes. In this case, the persistent notification indicates that the connection has dropped.

I have already disabled the battery optimization in my android 6.0. It could also be a real crash, I guess I should check logcat.

Device is moto g4 plus on stock ROM.

**Comment by mhaas** [comment]
This morning, I disabled my smart, location aware wifi enabling app.. dns66 has been running stable for about 16 hours. It looks like the other app was doing something funky to the network.

The other change I made was enabling adb debug, but that is probably not related.

**Comment by mhaas** [comment]
Looks like I spoke too soo. This morning, dns66 was rather unstable in my home wifi. Via 3g it seems stable. Not sure about the wifi at work, I don't use the device that much there.
I do have ipv6 at home, but not via 3g or at work.  Maybe that's related.

**Comment by julian-klode** [comment]
I can confirm this in OnePlus' build of Nougat. I don't think I saw that on Marshmallow, though.

**Comment by skoal3131** [comment]
Noob question.. on you dns filters does red mean its blocking ads or do i have to manually set to green. Also what does ignore do (grey) circle

**Comment by julian-klode** [comment]
@skoal3131 On the DNS list, you can configure custom DNS servers. There are actually only two values: On/Off, so red and grey both mean: Do not use this DNS server. It does not affect ad blocking. On the domain filter list, red means block, grey means ignore the entry (so you can temporarily disable stuff), and green allows you to whitelist stuff (at the end, last entry wins).

But to get back to the bug: I fixed a reconnecting error with Firefox (#48) in 73bafb568b5ab82eb00f9ab4bff8c284b3bfde3d - This might be related.

**Comment by MaluNoPeleke** [comment]
I have a similar problem on my OnePlus 3T with OOS 3.5.4:
DNS66 0.2.0 stopped working (no notification, no VPN key icon) and I had to start it again manually.

**Comment by julian-klode** [comment]
@MaluNoPeleke See #51 for the 0.2.0 issue, it's likely what you see.

**Comment by mhaas** [comment]
Seems to be fixed for me with 0.2.0. thanks a lot!

**Comment by Bebef** [comment]
I too have the problem that blocking sometimes stops working. When that happens and I try to stop and start the VPN in the app the app will report that there are no filter lists and I have to download them first.

Device Pixel, Android 7.1.1, DNS66 0.3.0

**Comment by smichel17** [comment]
I can reliably reproduce; I will turn off Doze and see if that changes my ability to reproduce.

**Comment by smichel17** [comment]
I turned off battery optimization for dns66 and I have not been able to reproduce in the last few days. Going to turn it back on again now.

**Comment by smichel17** [comment]
@julian-klode I'm pretty sure one of android's battery optimizations is the culprit here.

There is a way to detect whether your app is battery optimized or not. I don't know how, but DAVdroid does it.

**Comment by skoal3131** [comment]
Is this included in custom ROMs .. I finnaly figured twerp out on my g4 play but I think I still hate google and gapps. Ridiculous when I didn't have root. I like the unicorn ROM or whatever it is best . I take it lineage is kinda the base os. Sorry for all this nweb talk but I'm really googled out.....can I use that speech to to feature without google app. So tired of her asking me if I said something.. It unnerves me to kno she is ruining my songs by tryingv to speak over.

**Comment by julian-klode** [comment]
OK, in the linked commit 6a1861aad7c92ee54d172f3a5edcdd83d84a340c there are two dialogs: First the user is informed that power optimisations may cause issues, if he clicks yes, then a system-provided dialog pops up asking the user if he wants to exempt DNS66 from power optimisation.

**Comment by julian-klode** [comment]
The message is shown only once, and only if DNS66 is set to optimized.

**Comment by julian-klode** [comment]
Forgot to merge the ignore power optimization thingy :(

**Comment by nithinkolekar** [comment]
Issue still present with Android 5.1.1 of moto x pure. Attaching the screen shot where u can see VPN is not active and service is not started but  slide down bar show dns66 is active.
![screenshot_2017-07-18-20-22-46](https://user-images.githubusercontent.com/13523715/28323705-32ba4e46-6bf7-11e7-8727-09ff77fab39e.png)

**Comment by julian-klode** [comment]
That's exactly the opposite of what this bug describes.

**Comment by nithinkolekar** [comment]
What I meant was ,it was running and working properly for time before VPN and dns66 service  shutdowns silently but in slide down it is showing as running. There is no power optimization in Android 5.1.1 and mobother VPN app is enabled.
This happened just now in 45min of time frame. attaching the log 
[logcat-motox-dns66.txt](https://github.com/julian-klode/dns66/files/1158188/logcat-motox-dns66.txt)

**Comment by julian-klode** [comment]
What I meant is that the notification looks more accurate than the activity. But perhaps I'm wrong.

**Comment by nithinkolekar** [comment]
Any chance of identifying cause of service getting stopped  from the prev log or u need more bigger log?

**Comment by julian-klode** [comment]
It only says that Android told DNS66 that network connectivity changed to no connectivity. It obviously stops then, waiting for network.

**Comment by nithinkolekar** [comment]
But when network reconnects it should start service right? But it isn't and hence there is no log for service start activity I think.
Would you suggest any app that logs the network loss/reconnection, to further debug this issue?

**Comment by Karsty** [comment]
My phone has "background task clear" functionality listed under "settings".  To get DNS66 to stay running I had to flip the switch to add DNS66 to the whitelist.

**Comment by alleluia20** [comment]
I have the notification enabled at DNS66. I also have the "watch connection" option enabled. I have disabled, at Android (7.0), the battery saving for DNS66.
Still, DNS66 stops (or is killed) every few hours (sometimes, as little as two hours). It does not happen to Blokada (but the latter has its own problems :-)).
If you can imagine some information I can provide in order to debug this, please let me know...

**Comment by luigimannoni** [comment]
The issue is still present on Android Pie on a OnePlus 6T using version 0.5.8.

The service stops randomly and can't pinpoint to what's the exact cause, seems to be more noticeable when switching between networks (for example, more than often I notice it when coming out of the subway)

I do have the "watch connection" option enabled already and also took DNS66 off from the battery optimization list.
